File: 310971/Rogach/l-labs/service/src/main/java/com/labs/service/service/WriterService.java
package com.labs.service.service;

import com.labs.domain.entity.Writer;
import com.labs.domain.repository.WriterRepository;
import com.labs.service.dto.WriterDto;
import com.labs.service.exception.ResourceNotFoundException;
import com.labs.service.exception.ValidationException;
import com.labs.service.mapper.WriterMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
@Transactional
public class WriterService {
    private final WriterRepository writerRepository;
    private final WriterMapper writerMapper;

    public WriterDto create(WriterDto writerDto) {
        validateWriterDto(writerDto);
        if (writerRepository.findByLogin(writerDto.getLogin()).isPresent()) {
            throw new ValidationException("Writer with login '" + writerDto.getLogin() + "' already exists");
        }
        Writer writer = writerMapper.toEntity(writerDto);
        Writer saved = writerRepository.save(writer);
        return writerMapper.toDto(saved);
    }

    @Transactional(readOnly = true)
    public List<WriterDto> findAll() {
        return writerRepository.findAll().stream()
                .map(writerMapper::toDto)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public WriterDto findById(Long id) {
        Writer writer = writerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Writer with id " + id + " not found"));
        return writerMapper.toDto(writer);
    }

    public WriterDto update(Long id, WriterDto writerDto) {
        validateWriterDto(writerDto);
        Writer writer = writerRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Writer with id " + id + " not found"));
        
        if (!writer.getLogin().equals(writerDto.getLogin())) {
            if (writerRepository.findByLogin(writerDto.getLogin()).isPresent()) {
                throw new ValidationException("Writer with login '" + writerDto.getLogin() + "' already exists");
            }
        }
        
        writer.setLogin(writerDto.getLogin());
        writer.setPassword(writerDto.getPassword());
        writer.setFirstname(writerDto.getFirstname());
        writer.setLastname(writerDto.getLastname());
        
        Writer updated = writerRepository.save(writer);
        return writerMapper.toDto(updated);
    }

    public void delete(Long id) {
        if (!writerRepository.existsById(id)) {
            throw new ResourceNotFoundException("Writer with id " + id + " not found");
        }
        writerRepository.deleteById(id);
    }

    private void validateWriterDto(WriterDto writerDto) {
        if (writerDto == null) {
            throw new ValidationException("Writer data cannot be null");
        }
        if (writerDto.getLogin() == null || writerDto.getLogin().trim().isEmpty()) {
            throw new ValidationException("Login cannot be null or empty");
        }
        if (writerDto.getLogin().length() < 2 || writerDto.getLogin().length() > 64) {
            throw new ValidationException("Login must be between 2 and 64 characters");
        }
        if (writerDto.getPassword() == null || writerDto.getPassword().trim().isEmpty()) {
            throw new ValidationException("Password cannot be null or empty");
        }
        if (writerDto.getPassword().length() < 8 || writerDto.getPassword().length() > 128) {
            throw new ValidationException("Password must be between 8 and 128 characters");
        }
        if (writerDto.getFirstname() == null || writerDto.getFirstname().trim().isEmpty()) {
            throw new ValidationException("Firstname cannot be null or empty");
        }
        if (writerDto.getFirstname().length() < 2 || writerDto.getFirstname().length() > 64) {
            throw new ValidationException("Firstname must be between 2 and 64 characters");
        }
        if (writerDto.getLastname() == null || writerDto.getLastname().trim().isEmpty()) {
            throw new ValidationException("Lastname cannot be null or empty");
        }
        if (writerDto.getLastname().length() < 2 || writerDto.getLastname().length() > 64) {
            throw new ValidationException("Lastname must be between 2 and 64 characters");
        }
    }
}

File: 351002/Bubich/t330/publisher/app/services/writer_service.py
from typing import Dict, Any
from app.repository.database_repository import DatabaseRepository
from publisher.app.exceptions import NotFoundError, ValidationError


class WriterService:
    def __init__(self, repository: DatabaseRepository):
        self.repository = repository

    def create(self, data: Dict[str, Any]):
        self._validate_create(data)
        return self.repository.save(data)

    def get_by_id(self, id: int):
        writer = self.repository.find_by_id(id)
        if not writer:
            raise NotFoundError(f"Writer with id {id} not found")
        return writer

    def get_all(self):
        return self.repository.find_all()

    def update(self, id: int, data: Dict[str, Any]):
        self._validate_update(data)
        existing_writer = self.repository.find_by_id(id)
        if not existing_writer:
            raise NotFoundError(f"Writer with id {id} not found")

        update_data = {}
        if 'login' in data:
            update_data['login'] = data['login']
        if 'password' in data:
            update_data['password'] = data['password']
        if 'firstname' in data:
            update_data['firstname'] = data['firstname']
        if 'lastname' in data:
            update_data['lastname'] = data['lastname']

        return self.repository.update(id, update_data)

    def delete(self, id: int):
        """Удаление writer с проверкой существования"""
        writer = self.repository.find_by_id(id)
        if not writer:
            raise NotFoundError(f"Writer with id {id} not found")

        try:
            self.repository.delete_by_id(id)
        except Exception as e:
            raise Exception(f"Error deleting writer: {str(e)}")

    def _validate_create(self, data: Dict[str, Any]):
        if not data.get('login') or len(data['login']) < 2 or len(data['login']) > 64:
            raise ValidationError("Login must be between 2 and 64 characters", "40002")
        if not data.get('password') or len(data['password']) < 8 or len(data['password']) > 128:
            raise ValidationError("Password must be between 8 and 128 characters", "40003")
        if not data.get('firstname') or len(data['firstname']) < 2 or len(data['firstname']) > 64:
            raise ValidationError("Firstname must be between 2 and 64 characters", "40004")
        if not data.get('lastname') or len(data['lastname']) < 2 or len(data['lastname']) > 64:
            raise ValidationError("Lastname must be between 2 and 64 characters", "40005")

    def _validate_update(self, data: Dict[str, Any]):
        if 'login' in data and (len(data['login']) < 2 or len(data['login']) > 64):
            raise ValidationError("Login must be between 2 and 64 characters", "40002")
        if 'password' in data and (len(data['password']) < 8 or len(data['password']) > 128):
            raise ValidationError("Password must be between 8 and 128 characters", "40003")
        if 'firstname' in data and (len(data['firstname']) < 2 or len(data['firstname']) > 64):
            raise ValidationError("Firstname must be between 2 and 64 characters", "40004")
        if 'lastname' in data and (len(data['lastname']) < 2 or len(data['lastname']) > 64):
            raise ValidationError("Lastname must be between 2 and 64 characters", "40005")